File: desktop/mcp.ts
/**
 * The point of the desktop app: an MCP server on the diagram you are looking at.
 *
 * The editor already has an assistant, and it works — but it answers to a
 * Gemini key you pay for on top of the Claude Code, Codex or Antigravity
 * subscription you already have. This server is the same four tools offered to
 * the CLI you are already paying for, driving the same document through the
 * same loader.
 *
 * **The tools act on the open window, not on a file.** So they work on a
 * diagram that has never been saved, they land in the same undo stack as a
 * change made by hand, and the result is on screen before the tool returns.
 * The other route — a CLI that simply writes the `.arch.json` — is covered by
 * the file watcher, and needs nothing here.
 *
 * The names and the semantics are deliberately the assistant's
 * (`src/core/assistant.js`): a document written by Claude Code should meet the
 * same rules, and the same complaints, as one written by the panel.
 */

import { createMcpHandler, McpServer } from '@modelcontextprotocol/server';
import * as z from 'zod/v4';

type Bridge = { ask(name: string, args?: Record<string, unknown>): Promise<unknown> };

/** The port a CLI is told to use. Not sacred, but it has to be guessable. */
export const DEFAULT_MCP_PORT = 7337;

/**
 * Loopback is not access control.
 *
 * Any page in any browser on this machine can POST to `127.0.0.1:7337`, and
 * without this a visited web page could drive the editor — the DNS-rebinding
 * shape, where a hostile name resolves to loopback and the browser sends the
 * request for it. MCP clients send no `Origin` at all, so the rule is: no
 * origin is fine, a browser origin is not.
 */
function fromABrowser(req: Request): boolean {
  const origin = req.headers.get('origin');
  if (!origin) return false;
  try {
    const { hostname } = new URL(origin);
    return hostname !== '127.0.0.1' && hostname !== 'localhost' && hostname !== '[::1]';
  } catch {
    return true;
  }
}

const text = (body: string) => ({ content: [{ type: 'text' as const, text: body }] });

function buildServer(bridge: Bridge, version: string) {
  const server = new McpServer({ name: 'massing', version });

  const call = async (name: string, args?: Record<string, unknown>) => {
    try {
      return text(String(await bridge.ask(name, args)));
    } catch (err) {
      // Reported as content rather than thrown: a model that hears "the window
      // is not open" can say so, where a transport error is something it can
      // only retry.
      return text(`Refused: ${(err as Error).message}`);
    }
  };

  server.registerTool('get_diagram', {
    description:
      'Read the diagram open in the Massing editor, as a .arch.json document. Call this ' +
      'before any edit so you are changing what is actually on screen, and so you keep the ' +
      'ids that already exist.',
    inputSchema: z.object({}),
  }, () => call('get_diagram'));

  server.registerTool('replace_diagram', {
    description:
      'Replace what is open in the editor with a complete .arch.json document. Send the ' +
      'whole document, never a fragment: anything left out is deleted. The result reports ' +
      'whatever the loader had to repair, so read it and fix what it names rather than ' +
      'assuming the edit landed as written. The change is undoable in the editor.',
    inputSchema: z.object({
      document: z.string().describe('A complete .arch.json document as JSON text.'),
    }),
  }, ({ document }: { document: string }) => call('replace_diagram', { document }));

  server.registerTool(
    'add_tab',
    {
      description:
        'Add a new drawing to the open file as a tab, beside the one already open, and switch ' +
        'to it. Use it when one picture genuinely will not hold the answer — a system past ' +
        'about 25 blocks, or a second view the person asked for alongside the first. Never use ' +
        'it to tidy one diagram into several, and never to avoid editing what is open. You ' +
        'cannot create files: the file belongs to the person.',
      inputSchema: z.object({
        name: z.string().describe('What the tab is called, after what it shows.'),
        document: z.string().describe('The new drawing as a complete .arch.json document.'),
      }),
    },
    ({ name, document }: { name: string; document: string }) => call('add_tab', { name, document }),
  );

  server.registerTool('validate_diagram', {
    description:
      'Check the diagram on screen for the faults that make one unreadable: a block hidden ' +
      'behind a taller one, a connection that vanishes under a block, blocks overlapping, a ' +
      'block outside the zone it claims, captions written as sentences, too many connections ' +
      'for the number of blocks. Call it after every edit — none of this is visible in the ' +
      'JSON. ERROR means the picture is visibly broken. Act on the report rather than ' +
      'summarising it back.',
    inputSchema: z.object({}),
  }, () => call('validate_diagram'));

  return server;
}

/**
 * Start the server, and say where it actually landed.
 *
 * The port is asked for rather than assumed: 7337 may be taken by another copy
 * of this app or by something else entirely, and a desktop app that refuses to
 * start because a port is busy is a bad desktop app. The real port is written
 * where the setup instructions can find it.
 */
export function startMcp(bridge: Bridge, { version = '0.0.0', port = DEFAULT_MCP_PORT } = {}) {
  const handler = createMcpHandler(() => buildServer(bridge, version));

  const serve = (on: number) =>
    Deno.serve({
      hostname: '127.0.0.1',
      port: on,
      onError: (err) => {
        console.error('massing mcp:', err);
        return new Response('internal error', { status: 500 });
      },
    }, (req) => {
      if (fromABrowser(req)) {
        return new Response('Refused: this endpoint is not for browsers.', { status: 403 });
      }
      return handler.fetch(req);
    });

  let server: Deno.HttpServer<Deno.NetAddr>;
  try {
    server = serve(port);
  } catch (err) {
    if (!(err instanceof Deno.errors.AddrInUse)) throw err;
    console.error(`massing mcp: ${port} is taken, taking whatever is free instead`);
    server = serve(0);
  }

  const actual = server.addr.port;
  return {
    port: actual,
    url: `http://127.0.0.1:${actual}/`,
    stop: () => server.shutdown(),
  };
}

File: desktop/src-tauri/src/mcp.rs
//! The point of the desktop app: an MCP server on the diagram you are looking at.
//!
//! The editor has an assistant and it works, but it answers to a Gemini key you
//! pay for on top of the Claude Code, Codex or Antigravity subscription you
//! already have. This is the same four tools offered to the CLI you are already
//! paying for, driving the same document.
//!
//! **The tools act on the open window, not on a file.** So they work on a
//! diagram that has never been saved, they land in the same undo stack as a
//! change made by hand, and the result is on screen before the tool returns.
//! Each one is a round trip through `Bridge::ask` into the page, where the
//! implementations sit beside the editor's own loader — the same
//! `normalizeDoc` and `validateDocument` the panel uses, so a document written
//! by Claude Code meets the same rules and the same complaints.

use std::sync::Arc;

use rmcp::handler::server::router::tool::ToolRouter;
use rmcp::handler::server::wrapper::Parameters;
use rmcp::model::{CallToolResult, ContentBlock, Implementation, ServerCapabilities, ServerInfo};
use rmcp::transport::streamable_http_server::{
    session::local::LocalSessionManager, StreamableHttpServerConfig, StreamableHttpService,
};
use rmcp::{tool, tool_handler, tool_router, ErrorData as McpError, ServerHandler};
use serde_json::json;

use crate::bridge::Bridge;

/// The port a CLI is told to use. Not sacred, but it has to be guessable.
pub const DEFAULT_PORT: u16 = 7337;

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct ReplaceArgs {
    /// A complete .arch.json document as JSON text.
    pub document: String,
}

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct AddTabArgs {
    /// What the tab is called, after what it shows.
    pub name: String,
    /// The new drawing as a complete .arch.json document.
    pub document: String,
}

#[derive(Clone)]
pub struct Massing {
    bridge: Arc<Bridge>,
    /// Built by `#[tool_router]` and read by `#[tool_handler]`, both of which
    /// expand after the lint runs — so it looks unread and is not.
    #[allow(dead_code)]
    tool_router: ToolRouter<Massing>,
}

#[tool_router]
impl Massing {
    pub fn new(bridge: Arc<Bridge>) -> Self {
        Self {
            bridge,
            tool_router: Self::tool_router(),
        }
    }

    /// Every tool is the same shape: ask the window, hand back what it said.
    ///
    /// A refusal comes back as content rather than as an error, because a model
    /// that hears "the window is not open" can say so, where a transport
    /// failure is something it can only retry.
    async fn ask(&self, name: &str, args: serde_json::Value) -> Result<CallToolResult, McpError> {
        let said = match self.bridge.ask(name, args).await {
            Ok(answer) => answer,
            Err(why) => format!("Refused: {why}"),
        };
        Ok(CallToolResult::success(vec![ContentBlock::text(said)]))
    }

    #[tool(
        description = "Read the diagram open in the Massing editor, as a .arch.json document. \
        Call this before any edit so you are changing what is actually on screen, and so you \
        keep the ids that already exist."
    )]
    async fn get_diagram(&self) -> Result<CallToolResult, McpError> {
        self.ask("get_diagram", json!({})).await
    }

    #[tool(
        description = "Replace what is open in the editor with a complete .arch.json document. \
        Send the whole document, never a fragment: anything left out is deleted. The result \
        reports whatever the loader had to repair, so read it and fix what it names rather than \
        assuming the edit landed as written. The change is undoable in the editor."
    )]
    async fn replace_diagram(
        &self,
        Parameters(ReplaceArgs { document }): Parameters<ReplaceArgs>,
    ) -> Result<CallToolResult, McpError> {
        self.ask("replace_diagram", json!({ "document": document }))
            .await
    }

    #[tool(
        description = "Add a new drawing to the open file as a tab, beside the one already open, \
        and switch to it. Use it when one picture genuinely will not hold the answer — a system \
        past about 25 blocks, or a second view the person asked for alongside the first. Never \
        use it to tidy one diagram into several, and never to avoid editing what is open. You \
        cannot create files: the file belongs to the person."
    )]
    async fn add_tab(
        &self,
        Parameters(AddTabArgs { name, document }): Parameters<AddTabArgs>,
    ) -> Result<CallToolResult, McpError> {
        self.ask("add_tab", json!({ "name": name, "document": document }))
            .await
    }

    #[tool(
        description = "Check the diagram on screen for the faults that make one unreadable: a \
        block hidden behind a taller one, a connection that vanishes under a block, blocks \
        overlapping, a block outside the zone it claims, captions written as sentences, too many \
        connections for the number of blocks. Call it after every edit — none of this is visible \
        in the JSON. ERROR means the picture is visibly broken. Act on the report rather than \
        summarising it back."
    )]
    async fn validate_diagram(&self) -> Result<CallToolResult, McpError> {
        self.ask("validate_diagram", json!({})).await
    }
}

#[tool_handler]
impl ServerHandler for Massing {
    fn get_info(&self) -> ServerInfo {
        // `from_build_env` reads the *SDK's* crate name, so a client asking
        // who it is talking to hears "rmcp". Corrected by hand; the struct is
        // non-exhaustive, so the fields are set rather than constructed.
        let mut who = Implementation::from_build_env();
        who.name = "massing".into();
        who.version = env!("CARGO_PKG_VERSION").into();
        ServerInfo::new(ServerCapabilities::builder().enable_tools().build()).with_server_info(who)
    }
}

/// Build the tower service the HTTP server mounts.
pub fn service(bridge: Arc<Bridge>) -> StreamableHttpService<Massing, LocalSessionManager> {
    StreamableHttpService::new(
        move || Ok(Massing::new(Arc::clone(&bridge))),
        LocalSessionManager::default().into(),
        StreamableHttpServerConfig::default(),
    )
}
